Cut v2.50.0-rc.0 (#13465)
prometheus/prometheus: CHANGELOG.md

# Changelog

## 2.50.0-rc.0 / 2024-02-07

> * [CHANGE] Remote Write: Drop old inmemory samples. Activated using the config entry `sample_age_limit` #13002

Reviewer: This feels more like a feature than a change.

> * [FEATURE] Remote Write: Drop old inmemory samples. Activated using the config entry `sample_age_limit` #13002

Reviewer 2: Usually there are periods at the end of entries before the PR references

```suggestion
* [FEATURE] Remote Write: Drop old inmemory samples. Activated using the config entry `sample_age_limit`. #13002
```

* [FEATURE] **Experimental**: Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion`). #12733 #13279

> * [FEATURE] Promtool: Add `analyze` histograms command #12331

Reviewer 2: ```suggestion
* [FEATURE] Promtool: Add `analyze` histograms command. #12331
```

> * [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction #13282 #13393 #13398

Reviewer 2: ```suggestion
* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction. #13282 #13393 #13398
```

> * [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit` #13395

Reviewer 2: ```suggestion
* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit`. #13395
```

> * [ENHANCEMENT] Promtool: allow setting multiple matchers to `promtool tsdb dump`. #13296

Reviewer 2: While the above is what the PR title says, I found it a little confusing, especially the use of "to" ("setting" vs "specifying" doesn't really matter)

```suggestion
* [ENHANCEMENT] Promtool: allow specifying multiple matchers in `promtool tsdb dump`. #13296
```

* [ENHANCEMENT] PromQL: Restore more efficient version of `NewPossibleNonCounterInfo` annotation. #13022

> * [ENHANCEMENT] Kuma SD: Extend configuration to allow users to specify ClientId. #13278

Reviewer 2: Since "ClientID" isn't even the actual field name that got added here this might make more sense.

```suggestion
* [ENHANCEMENT] Kuma SD: Extend configuration to allow users to specify a client ID. #13278
```

* [ENHANCEMENT] PromQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. This is **experimental**. #13411
* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222

> * [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392

Reviewer: ```suggestion
* [ENHANCEMENT] Native Histograms: Issue warning if histogramRate is applied to the wrong kind of histogram. #13392
```

(this is the title of the issue, which seems more meaningful to me)

* [ENHANCEMENT] TSDB: Make transaction isolation data structures smaller. #13015
* [ENHANCEMENT] TSDB/postings: Optimize merge using Loser Tree. #12878

> * [ENHANCEMENT] TSDB: Simplify internal series delete function. #13261

Reviewer 2: This seems like more of an internal thing?

Reviewer: It's a small optimization.

> * [ENHANCEMENT] AGENT: Performance improvement by making the global hash lookup table smaller. #13262

Reviewer 2: ```suggestion
* [ENHANCEMENT] Agent: Performance improvement by making the global hash lookup table smaller. #13262
```

> * [ENHANCEMENT] Labels: Add DropMetricName function, used in PromQL. #13446

Reviewer: ```suggestion
* [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate() #13446
```

(citing the benefit rather than the low-level implementation)

> * [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate() #13446

Reviewer 2: ```suggestion
* [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate(). #13446
```

* [ENHANCEMENT] TSDB: Optimize label values with matchers by taking shortcuts. #13426
* [ENHANCEMENT] Kubernetes SD: Check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
* [ENHANCEMENT] Promtool: Improve visibility for `promtool test rules` with JSON colored formatting. #13342
* [ENHANCEMENT] Consoles: Exclude iowait and steal from CPU Utilisation. #9593

> * [ENHANCEMENT] Various improvements and optimizations on Native Histogram. #13267, #13215, #13276 #13289, #13340

Reviewer 2: ```suggestion
* [ENHANCEMENT] Various improvements and optimizations on Native Histograms. #13267, #13215, #13276 #13289, #13340
```

> * [BUGFIX] Fix quality value in HTTP Accept header. #13313

Reviewer 2: ```suggestion
* [BUGFIX] Scraping: Fix quality value in HTTP Accept header. #13313
```

# Changelog

## 2.50.0-rc.0 / 2024-01-26

> * [FEATURE] Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion`). #12733 #13279

Author: isn't obvious since it's behind a feature-flag ?

> * [FEATURE] Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion`). #12733 #13279

Author: done

> * [FEATURE] TSDB/{index,compact}: allow using custom postings encoding format. #13242 #13407 #13391

Author: ok I will remove it. But "user" can mean so many different profiles. Here it could be a developer that would still be interested about this feature don't you think ?

* [FEATURE] Promtool: Add `analyze` histograms command #12331
* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction #13282 #13393 #13398
* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit` #13395
* [ENHANCEMENT] Promtool: allow setting multiple matchers to `promtool tsdb dump`. #13296
* [ENHANCEMENT] PromQL: Restore more efficient version of `NewPossibleNonCounterInfo` annotation. #13022
* [ENHANCEMENT] Enable reusing memory when converting between histogram types. #13215

> * [ENHANCEMENT] PERFORMANCE: Reuse float histogram objects. #13276 #13289

Author: Yeah I know but there are so many PRs relative to the performance, could be interested to introduce it

> * [ENHANCEMENT] SD: Extend Kuma SD configuration to allow users to specify ClientId. #13278

Reviewer 3: ```suggestion
* [ENHANCEMENT] Kuma SD: Extend configuration to allow users to specify ClientId. #13278
```

* [ENHANCEMENT] Optimize histogram iterators. #13340

> * [ENHANCEMENT] PROMQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. #13411

Reviewer: Suggest to add 'experimental' on this line.

* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
* [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392

> * [ENHANCEMENT] TSDB/PERFORMANCE: make transaction isolation data structures smaller. #13015

Reviewer 3: ```suggestion
* [ENHANCEMENT] TSDB: Make transaction isolation data structures smaller. #13015
```

> * [ENHANCEMENT] TSDB/postings: use Loser Tree for merge. #12878

Reviewer: ```suggestion
* [ENHANCEMENT] TSDB/postings: optimize merge using Loser Tree. #12878
```

> * [ENHANCEMENT] TSDB/PERFORMANCE: simplify internal series delete function. #13261

Reviewer 3: ```suggestion
* [ENHANCEMENT] TSDB: simplify internal series delete function. #13261
```

> * [ENHANCEMENT] TSDB/AGENT: Performance improvement by making the global hash lookup table smaller. #13262

Reviewer 3: ```suggestion
* [ENHANCEMENT] Agent: Performance improvement by making the global hash lookup table smaller. #13262
```

> * [ENHANCEMENT] Labels/PERFORMANCE: Add DropMetricName function, used in PromQL. #13446

Reviewer 3: we do not need to mention this

Reviewer: We could mention it as an optimization to PromQL expressions like `abs(foo)`.
It's also possibly a pessimization, since we no longer cache the output.  If people have labels like `__i_sort_before__name`.

Author: well either way is fine for me.
Perhaps I should rephrase it to make it more user oriented like @bboreham is suggesting ?

Author: Would you be fine with that @roidelapluie ?

> * [ENHANCEMENT] PERFORMANCE: Optimize label values with matchers by taking shortcuts. #13426

Reviewer 3: ```suggestion
* [ENHANCEMENT] TSDB: Optimize label values with matchers by taking shortcuts. #13426
```

> * [ENHANCEMENT] SD/PERFORMANCE: check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408

Reviewer 3: ```suggestion
* [ENHANCEMENT] Kubernetes SD: check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
```

> * [ENHANCEMENT] PromQL: simplify Native Histogram arithmetics. #13267

Reviewer: Could bundle together various native histogram optimisations?  #13267, #13215, #13276 #13289, #13340

Author: yeah that will be more consistant.

* [ENHANCEMENT] Promtool: Improve visibility for `promtool test rules` with JSON colored formatting. #13342

> * [ENHANCEMENT] CONSOLES: exclude iowait and steal from CPU Utilisation. #9593

Reviewer 3: ```suggestion
* [ENHANCEMENT] Consoles: Exclude iowait and steal from CPU Utilisation. #9593
```

* [BUGFIX] Fix quality value in HTTP Accept header. #13313
* [BUGFIX] Fix reusing float histograms. #13289

> * [BUGFIX] UI: fix handling of scalar and string in isHeatmapData. #13371

Reviewer: Can we make this more end-user oriented?  When would they notice the difference?

Author: done. Let me know if you think it is fuzzy

> * [CHANGE] REMOTE_WRITE: Drop old inmemory samples. Activated using the config entry `sample_age_limit` #13002

Reviewer 3: ```suggestion
* [CHANGE] Remote Write: Drop old inmemory samples. Activated using the config entry `sample_age_limit` #13002
```

Reviewer: Is this not an enhancement?  CHANGE to me means a breaking change I get without any config change.

Author: I think when I wrote this, I wanted to stamp, it's a change when you are activating it. I'm even wondering if you can rollback after activating it. It's probably the case.

> * [CHANGE] REMOTE_WRITE: Error `storage.ErrTooOldSample` is now generating HTTP error 400 instead of HTTP error 500. #13335

Reviewer 3: ```suggestion
* [CHANGE] Remote Write: Error `storage.ErrTooOldSample` is now generating HTTP error 400 instead of HTTP error 500. #13335
```

Reviewer: I would put breaking changes at the top.

# Changelog

## 2.50.0-rc.0 / 2024-01-26

> * [FEATURE] Add new feature flag `created-timestamp-zero-ingestion`. #12733 #13279

Reviewer 4: ```suggestion
* [FEATURE] Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion)`. #12733 #13279
```

* [FEATURE] TSDB/{index,compact}: allow using custom postings encoding format. #13242 #13407 #13391
* [FEATURE] Promtool: Add `analyze` histograms command #12331
* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction #13282 #13393 #13398

> * [FEATURE] Add automatic memory limit handling. Activated using the flag `auto-gomemlimit.ratio` #13395

Reviewer 5: ```suggestion
* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit` #13395
```

* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
* [ENHANCEMENT] Native Histograms: Issue warning if histogramRate is applied to the wrong kind of histogram. #13392
* [ENHANCEMENT] TSDB: Make transaction isolation data structures smaller. #13015
* [ENHANCEMENT] TSDB/postings: Optimize merge using Loser Tree. #12878
* [ENHANCEMENT] TSDB: Simplify internal series delete function. #13261
* [ENHANCEMENT] Agent: Performance improvement by making the global hash lookup table smaller. #13262
* [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate() #13446
* [ENHANCEMENT] TSDB: Optimize label values with matchers by taking shortcuts. #13426
* [ENHANCEMENT] Kubernetes SD: Check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
* [ENHANCEMENT] Promtool: Improve visibility for `promtool test rules` with JSON colored formatting. #13342
* [ENHANCEMENT] Consoles: Exclude iowait and steal from CPU Utilisation. #9593
* [ENHANCEMENT] Various improvements and optimizations on Native Histograms. #13267, #13215, #13276 #13289, #13340

> * [BUGFIX] Scrapping: Fix quality value in HTTP Accept header. #13313

Reviewer: ```suggestion
* [BUGFIX] Scraping: Fix quality value in HTTP Accept header. #13313
```
